DRMAAV2-20: Create connection pool for DRMAA to connect with DRMS(PBSPro)

rampranesh opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[DRMAAV2-20](https://pbspro.atlassian.net/browse/DRMAAV2-20)**

#### Problem description
* Add connection pool for DRMAA to connect with PBS.

#### Cause / Analysis
* NA

#### Solution description
* Added Connection pool implementation.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **unit test(s) to my commit**. *(or)*
- [ ] I have added **manual test(s) to the Jira ticket and explained why unit testing is not appropriate** for this case.
- [x] All new and existing automated tests have passed. 
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

vinodchitrali commented on inc/ConnectionPool.h:
```diff
@@ -0,0 +1,107 @@
+/**
+  *  @brief Class that maintains File descriptor, Port, Server name and value to indicate its status.  
+  */  
+class Connection {
```
U can use of Object pool pattern something like this 

https://en.wikipedia.org/wiki/Object_pool_pattern

vinodchitrali commented on inc/ConnectionPool.h:
```diff
@@ -0,0 +1,107 @@
+		 *
+		 */
+		ConnPool() {}
+		list<Connection> _connPool;
```
Use vector .. vector gives better performance than list

suresh-thelkar commented on src/ConnectionPool.cpp:
```diff
@@ -0,0 +1,179 @@
+	if(foundFd != true) {
+		pthread_mutex_unlock(&_instMutex);
+		return -1;
+	}
```
Before we call pbs_connect, it would be better if we call pbs_disconnect or the corresponding wrapper.

rampranesh replied:
Hi @suresh-thelkar ,
   I was looking to add the pbs_disconnect as you have suggested but in a multi-threaded application, we might end up closing a connection thinking that it is stale. But another thread would have got the same FD after reconnection. Please let me know your views on this.
Thanks,
Rampranesh M

suresh-thelkar replied:
As discussed, as per the new code which is checked in today pbs_disconnect() may not be needed as you are tagging thread id to the corresponding fd.

rampranesh commented:
Hi @vinodchitrali,
 I have added the changes which you have mentioned, please have a look at the changes.

Thanks,
Rampranesh M

vinodchitrali commented on inc/ConnectionPool.h:
```diff
@@ -0,0 +1,112 @@
+		int _pid;
+	public:
+		DRMSConnection() { _fd = 0; _port = 0; _pid = 0;}
+		DRMSConnection(int fd_, string serverName_, int port_, int pid_) { _fd = fd_; _port = port_; _pid = pid_; }
```
arguments should be const and reference

vinodchitrali commented on test/unittesting/src/Makefile.am:
```diff
@@ -36,9 +36,10 @@
+test_drmaa_SOURCES=	testdrmaa.cpp \
+			runtest.cpp
 						
 test_drmaa_LDADD = ../../../api/libdrmaav2.la -lcppunit
```
We haven't added cppunit license  info in License file. 
U can use this ticket to add license info

rampranesh commented:
Hi @vinodchitrali ,
  I have addressed your comments please have a look at these changes.

Thanks,
Rampranesh M

vinodchitrali commented:
Maintainers please merge this PR.

arungrover commented on .travis.yml:
```diff
@@ -0,0 +1,41 @@
+    - ${DOCKER_EXEC} zypper -n install sendmail
+    - ${DOCKER_EXEC} zypper -n install gcc-c++
+    - ${DOCKER_EXEC} zypper -n install cppunit-devel
+    - ${DOCKER_EXEC} lynx -dump http://pbspro.org/Download.aspx | awk '/http/{print $2}' | grep openSUSE > link
```
I don't really have a good solution to this, but this is going to fail if there were more than one openSUSE release present (like 17.1.0)
For now, I guess just search for OpenSUSE_13.2.zip instead of openSUSE. We should bring this up that instead of doing this search for link and then fetching the packages there should be a fileserver from where packages can be fetched.

arungrover commented on inc/ConnectionPool.h:
```diff
@@ -0,0 +1,277 @@
+		}
+		/**
+		 * @brief
+		 *	setFd() - sets connection fd with value passed
```
Please specify arguments too in doxygen header

rampranesh replied:
Done, added arguments to Doxygen in all methods.

arungrover commented on inc/ConnectionPool.h:
```diff
@@ -0,0 +1,277 @@
+		 *
+		 */
+		void* getDetails() {
+			int *_pfd = new int;
```
What is the use of having the data member private when we are giving away it's address itself.
What is the use of getDetails() method?

rampranesh replied:
This was added initially when the abstract class has been declared, now that we have other meaningful pure virtual functions in the class we don't need this anymore. So removed it.

arungrover commented on inc/ConnectionPool.h:
```diff
@@ -0,0 +1,277 @@
+		 * @brief
+		 *	connect() - Pure virtual function to connect to DRMS.
+		 */
+		virtual void connect() = 0;
```
Does it throw an exception? do you want to mention that in the declaration here?

rampranesh replied:
Done

arungrover commented on inc/ConnectionPool.h:
```diff
@@ -0,0 +1,277 @@
+		 *
+		 */
+		Connection* clone() const {
+			return new PBSConnection(*this);
```
This constructor does not really dup the fd rather it just makes a copy of it. This means that if we clone using this method and one of the instance is deleted then the fd value pointed by the cloned connection is also closed. right?

rampranesh replied:
Added dup call for file descriptor.

arungrover commented on inc/ConnectionPool.h:
```diff
@@ -0,0 +1,277 @@
+		 *
+		 */
+		void connect() {
+			_fd = pbs_connect((char *)_serverName.c_str()); //TODO: Will replace with wrapped DRMS System apis
```
Where is the destructor? when is this connection closed?

rampranesh replied:
I have added destructor to the class, We are not deleting the object once we are not using the object same will be added to another list and will be used throughout the SessionManager's lifetime. So if session manager wants to reset the connection pool, it was not possible before. So introduced clearConnectionPool method.

arungrover commented on inc/ConnectionPool.h:
```diff
@@ -0,0 +1,277 @@
+		 * @return    pointer to ConnectionPool object
+		 *
+		 */
+		static ConnectionPool* getInstance() {
```
If this is a singleton then why don't we use the same singleton template class as other classes are using. We shouldn't be having two different ways to making a class singleton.
make '=' operator as private too

rampranesh replied:
I am waiting for DRMAAV2-23 PR to get checked in so that I can add the same singleton template and all needed exceptions to the code.

arungrover replied:
Some part of the code is already in, Session Manager Impl is making use of that template class already. Operator still needs to be private

rampranesh commented:
Hi @arungrover ,
I have addressed your comments, also have removed the CI part of the code. Will raise it as a different PR. Could you please have a look at the changes.
Thanks,
Rampranesh M

arungrover commented on src/ConnectionPool.cpp:
```diff
@@ -0,0 +1,118 @@
+		Connection *cnHold_ = _freeConnections.front();
+		_usedConnections.push_back(cnHold_);
+		_freeConnections.pop_front();
+		pthread_mutex_unlock(&ConnectionPool::_connMutex);
```
Some of these functions like connect(), disconnect are going to throw exceptions too. I'd suggest that you make a locker class, create it's local object in each of these functions where you want to take a lock  (take a lock in locker's constructor and release the lock in its destructor)

rampranesh replied:
Hi @arungrover ,
   We are planning to implement it in later stage. As of now can I have it the way it is now so that once the locker class is defined I shall add the same.

Thanks,
Rampranesh M

rampranesh commented:
Hi @arungrover ,
   I have addresses your comments, could you please have a look at the changes.

Thanks,
Rampranesh M

arungrover commented:
looks good @rampranesh, once you have another reviewer sign-off, rebase and squash.

rampranesh commented:
Hi @arungrover ,
   I have squashed and rebased my commits, could you please merge this PR.

Thanks,
Rampranesh M

arungrover commented:
Merged
